File: solutions/elixir/rational-numbers/2/lib/rational_numbers.ex
defmodule RationalNumbers do
  @type rational :: {integer, integer}

  @doc """
  Add two rational numbers
  """
  @spec add(a :: rational, b :: rational) :: rational
  def add({a1, b1}, {a2, b2}),
    do: {a1 * b2 + a2 * b1, b1 * b2} |> reduce

  @doc """
  Subtract two rational numbers
  """
  @spec subtract(a :: rational, b :: rational) :: rational
  def subtract({a1, b1}, {a2, b2}),
    do: {a1 * b2 - a2 * b1, b1 * b2} |> reduce

  @doc """
  Multiply two rational numbers
  """
  @spec multiply(a :: rational, b :: rational) :: rational
  def multiply({a1, b1}, {a2, b2}),
    do: {a1 * a2, b1 * b2} |> reduce

  @doc """
  Divide two rational numbers
  """
  @spec divide_by(num :: rational, den :: rational) :: rational
  def divide_by({a1, b1}, {a2, b2}),
    do: {a1 * b2, a2 * b1} |> reduce

  @doc """
  Absolute value of a rational number
  """
  @spec abs(a :: rational) :: rational
  def abs({a, b}),
    do: {Kernel.abs(a), Kernel.abs(b)} |> reduce

  @doc """
  Exponentiation of a rational number by an integer
  """
  @spec pow_rational(a :: rational, n :: integer) :: rational
  def pow_rational({a, b}, n) when n < 0,
    do: {b ** -n, a ** -n} |> reduce

  def pow_rational({a, b}, n),
    do: {a ** n, b ** n} |> reduce

  @doc """
  Exponentiation of a real number by a rational number
  """
  @spec pow_real(x :: integer, n :: rational) :: float
  def pow_real(x, {a, b}),
    do: x ** (a / b)

  @doc """
  Reduce a rational number to its lowest terms
  """
  @spec reduce(a :: rational) :: rational
  def reduce({a, b}) when b < 0,
    do: {-a, -b} |> reduce

  def reduce({a, b}) do
    gcd = Integer.gcd(a, b)
    {div(a, gcd), div(b, gcd)}
  end
end

File: solutions/lua/rational-numbers/2/rational-numbers.lua
local function gcd(a, b) return b == 0 and a or gcd(b, a % b) end

local function reduce(a)
	local gcf = gcd(a[1], a[2])
	return { a[1] // gcf, a[2] // gcf }
end

local function add(a, b) return reduce({ a[1] * b[2] + b[1] * a[2], a[2] * b[2] }) end
local function subtract(a, b) return reduce({ a[1] * b[2] - b[1] * a[2], a[2] * b[2] }) end
local function multiply(a, b) return reduce({ a[1] * b[1], a[2] * b[2] }) end
local function divide(a, b) return reduce({ a[1] * b[2], a[2] * b[1] }) end
local function abs(a) return reduce({ math.abs(a[1]), math.abs(a[2]) }) end
local function exp_rational(a, p) return p < 0 and reduce({ a[2] ^ -p, a[1] ^ -p }) or reduce({ a[1] ^ p, a[2] ^ p }) end
local function exp_real(p, a) return p ^ (a[1] / a[2]) end

return {
	add = add,
	subtract = subtract,
	multiply = multiply,
	divide = divide,
	abs = abs,
	exp_rational = exp_rational,
	exp_real = exp_real,
	reduce = reduce,
}
